prime-router/src/main/kotlin/serializers/Hl7Serializer.kt

```diff
@@ -181,64 +185,74 @@ class Hl7Serializer(val metadata: Metadata) {
+                        val question = it.hl7AOEQuestion!!
+                        val countObservations = 10
+                        // todo: map each AOE by the AOE question ID
+                        for (c in 0 until countObservations) {
+                            var spec = "/.OBSERVATION($c)/OBX-3-1"
+                            val questionCode = try {
+                                terser.get(spec)
+                            } catch (e: HL7Exception) {
+                                // todo: convert to result detail, maybe
+                                errors.add("Exception for $spec: ${e.message}")
+                                null
+                            }
+                            if (questionCode?.startsWith(question) == true) {
+                                spec = "/.OBSERVATION($c)/OBX-5"
+                                queryTerserForValue(terser, spec, it.name, mappedRows, errors, warnings)
+                            }
                         }
+                        "/.AOE"
+                    } else {
+                        "/.${it.hl7Field}"
                     }
-                    "/.AOE"
-                } else {
-                    "/.${it.hl7Field}"
-                }
 
-                if (terserSpec != "/.AOE") {
-                    queryTerserForValue(terser, terserSpec, it.name, mappedRows, errors, warnings)
-                } else {
-                    if (mappedRows[it.name]?.isEmpty() == true) mappedRows[it.name]?.add("")
-                }
-            } else {
-                it.hl7OutputFields?.forEach { h ->
-                    val terserSpec = if (h.startsWith("MSH")) {
-                        "/$h"
+                    if (terserSpec != "/.AOE") {
+                        queryTerserForValue(terser, terserSpec, it.name, mappedRows, errors, warnings)
                     } else {
-                        "/.$h"
+                        if (mappedRows[it.name]?.isEmpty() == true) mappedRows[it.name]?.add("")
+                    }
+                } else {
+                    it.hl7OutputFields?.forEach { h ->
+                        val terserSpec = if (h.startsWith("MSH")) {
+                            "/$h"
+                        } else {
+                            "/.$h"
+                        }
+                        queryTerserForValue(terser, terserSpec, it.name, mappedRows, errors, warnings)
                     }
-                    queryTerserForValue(terser, terserSpec, it.name, mappedRows, errors, warnings)
                 }
             }
+        } catch (e: Exception) {
+            errors.add("${e.localizedMessage} ${e.stackTraceToString()}")
         }
+
         // convert sets to lists
-        val rows = mappedRows.keys.map {
-            it to (mappedRows[it]?.toList() ?: emptyList())
-        }.toMap()
+        val rows = mappedRows.keys.associateWith {
+            (mappedRows[it]?.toList() ?: emptyList())
+        }
 
```

Comment:
`associateWith` is a nice piece of Kotlin magic that takes the `it` value and associates it with the return value from the closure and automagically creates a map out of it.

Reply:
A lovely gem!    I found a Kotlin gem recently too, 'groupingBy' that essentially does the same as a 'group by' in sql.